lib/mongoid/association/referenced/has_one/proxy.rb

```diff
@@ -49,20 +49,23 @@ def nullify
           #
           # @return [ One ] The association.
           def substitute(replacement)
-            # If the same object currently associated is being assigned,
-            # rebind the association and save the target but do not destroy
-            # the target.
+            if equal?(replacement)
+              self
+            else
+              unbind_one
+              if persistable?
+                if _association.destructive?
+                  send(_association.dependent)
+                else
+                  save if persisted?
+                end
+              end
 
-            unbind_one
-            if persistable?
-              # TODO can this entire method be skipped if self == replacement?
-              if _association.destructive? && self != replacement
-                send(_association.dependent)
-              else
-                save if persisted?
+              if replacement
+                reset_foreign_key!(replacement)
```

Comment:
This solves the problem of the stale foreign key not allowing the update to persist. It sets the foreign key to nil without updating the changed_attributes of the replacement document. This is okay, because inside the binding in the HasOne Proxy, the foreign key is set to the _base's key anyway. This will not affect the outcome of the operation at all, unless the replacement document has the same value for the foreign key as the original document which could only happen in the following three cases:
1. We are assigning the same document twice to the has_one association. This case doesn't make it through this code path, and exits early at the `equal?` condition above.
2. A second object created with the foreign key explicitly set.
3. Our case, where the foreign key is stale
The last two case are where our problem arises and this function solves that. In my opinion, those writes should still work.
There is one possible problem, however. There are multiple places where an error can be raised between the time that we set the foreign key to nil and the foreign key is set to the correct value. If this happens then we just altered the value of the foreign key on the object the user gave to us and the operation failed. Now, this operation was executed with the assumption that the foreign key would be changed, but setting it to nil is not what the user would expect. I'm not sure what the way forward is here.
Note, doing this operation right before we set the foreign key doesn't work, since the binding is used in many places that we don't want to reset the foreign key.